File: ide_extensions/vscode/test-files/vulnerable_sample.py
#!/usr/bin/env python3
"""
Sample vulnerable Python file for testing DeVAIC VS Code Extension
This file contains various security vulnerabilities for demonstration
"""

import os
import subprocess
import sqlite3
import random
from flask import Flask, request, render_template_string

app = Flask(__name__)

# 1. Hardcoded secret (CWE-798)
API_KEY = "sk-1234567890abcdef"
DATABASE_PASSWORD = "admin123"

# 2. SQL Injection vulnerability (CWE-89)
@app.route('/user/<user_id>')
def get_user(user_id):
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    # Vulnerable: Direct string concatenation
    query = f"SELECT * FROM users WHERE id = {user_id}"
    cursor.execute(query)
    
    result = cursor.fetchone()
    conn.close()
    return str(result)

# 3. Command Injection vulnerability (CWE-78)
@app.route('/execute')
def execute_command():
    cmd = request.args.get('cmd', '')
    
    # Vulnerable: Direct command execution
    result = subprocess.run(f"ls {cmd}", shell=True, capture_output=True)
    return result.stdout.decode()

# 4. Cross-Site Scripting (XSS) vulnerability (CWE-79)
@app.route('/greet')
def greet():
    name = request.args.get('name', 'Guest')
    
    # Vulnerable: Unescaped user input in template
    template = f"<h1>Hello {name}!</h1>"
    return render_template_string(template)

# 5. Weak random number generation (CWE-338)
def generate_session_token():
    # Vulnerable: Using predictable random
    return str(random.randint(1000000, 9999999))

# 6. Information disclosure (CWE-200)
@app.route('/debug')
def debug_info():
    # Vulnerable: Exposing sensitive system information
    return {
        'environment': dict(os.environ),
        'python_path': os.sys.path,
        'current_directory': os.getcwd(),
        'database_password': DATABASE_PASSWORD
    }

# 7. Path traversal vulnerability (CWE-22)
@app.route('/file')
def read_file():
    filename = request.args.get('file', '')
    
    # Vulnerable: No path validation
    try:
        with open(filename, 'r') as f:
            return f.read()
    except Exception as e:
        return str(e)

# 8. Insecure deserialization (CWE-502)
import pickle

@app.route('/load_data', methods=['POST'])
def load_data():
    data = request.get_data()
    
    # Vulnerable: Deserializing untrusted data
    obj = pickle.loads(data)
    return str(obj)

# 9. Missing authentication
@app.route('/admin/delete_user/<user_id>')
def delete_user(user_id):
    # Vulnerable: No authentication check
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    cursor.execute(f"DELETE FROM users WHERE id = {user_id}")
    conn.commit()
    conn.close()
    return "User deleted"

# 10. Weak cryptographic practices (CWE-327)
import hashlib

def hash_password(password):
    # Vulnerable: Using MD5 for password hashing
    return hashlib.md5(password.encode()).hexdigest()

# 11. Race condition (CWE-362)
import threading

balance = 1000
lock = threading.Lock()

def withdraw(amount):
    global balance
    # Vulnerable: No proper locking
    if balance >= amount:
        # Race condition possible here
        balance -= amount
        return True
    return False

# 12. Buffer overflow potential (in Python it's less common, but shown for demo)
def process_data(data):
    # Vulnerable: No length validation
    buffer = [0] * 1024
    
    # This could cause issues if data is too long
    for i, byte in enumerate(data):
        buffer[i] = byte
    
    return buffer

if __name__ == '__main__':
    # Vulnerable: Debug mode in production
    app.run(debug=True, host='0.0.0.0')

File: ide_extensions/vscode/test-files/vulnerable_sample.js
// Sample vulnerable JavaScript file for testing DeVAIC VS Code Extension
// This file contains various security vulnerabilities for demonstration

const express = require('express');
const mysql = require('mysql');
const crypto = require('crypto');
const { exec } = require('child_process');

const app = express();

// 1. Hardcoded secrets (CWE-798)
const API_KEY = 'abc123def456';
const DB_PASSWORD = 'password123';

// 2. SQL Injection vulnerability (CWE-89)
app.get('/user/:id', (req, res) => {
    const userId = req.params.id;
    
    // Vulnerable: String concatenation in SQL query
    const query = `SELECT * FROM users WHERE id = ${userId}`;
    
    connection.query(query, (error, results) => {
        if (error) throw error;
        res.json(results);
    });
});

// 3. Command Injection vulnerability (CWE-78)
app.get('/execute', (req, res) => {
    const command = req.query.cmd;
    
    // Vulnerable: Direct command execution
    exec(`ls ${command}`, (error, stdout, stderr) => {
        if (error) {
            res.status(500).send(error.message);
            return;
        }
        res.send(stdout);
    });
});

// 4. Cross-Site Scripting (XSS) vulnerability (CWE-79)
app.get('/search', (req, res) => {
    const query = req.query.q;
    
    // Vulnerable: Unescaped user input in HTML
    const html = `
        <html>
            <body>
                <h1>Search Results for: ${query}</h1>
                <div id="results"></div>
                <script>
                    document.getElementById('results').innerHTML = '${query}';
                </script>
            </body>
        </html>
    `;
    res.send(html);
});

// 5. Insecure randomness (CWE-338)
function generateToken() {
    // Vulnerable: Using Math.random() for security-sensitive operations
    return Math.random().toString(36).substring(2);
}

// 6. Information disclosure (CWE-200)
app.get('/debug', (req, res) => {
    // Vulnerable: Exposing sensitive information
    res.json({
        environment: process.env,
        nodeVersion: process.version,
        platform: process.platform,
        apiKey: API_KEY,
        dbPassword: DB_PASSWORD
    });
});

// 7. Prototype pollution (CWE-1321)
function merge(target, source) {
    for (let key in source) {
        // Vulnerable: No check for __proto__ or constructor
        if (typeof source[key] === 'object' && source[key] !== null) {
            if (!target[key]) target[key] = {};
            merge(target[key], source[key]);
        } else {
            target[key] = source[key];
        }
    }
    return target;
}

app.post('/merge', (req, res) => {
    const result = {};
    // Vulnerable: Merging untrusted user input
    merge(result, req.body);
    res.json(result);
});

// 8. Weak cryptographic practices (CWE-327)
function hashPassword(password) {
    // Vulnerable: Using MD5 for password hashing
    return crypto.createHash('md5').update(password).digest('hex');
}

// 9. Path traversal vulnerability (CWE-22)
app.get('/file', (req, res) => {
    const filename = req.query.file;
    
    // Vulnerable: No path validation
    const fs = require('fs');
    fs.readFile(filename, 'utf8', (err, data) => {
        if (err) {
            res.status(500).send(err.message);
            return;
        }
        res.send(data);
    });
});

// 10. Insecure direct object reference (CWE-639)
const users = {
    '1': { name: 'Admin', role: 'admin', ssn: '123-45-6789' },
    '2': { name: 'User', role: 'user', ssn: '987-65-4321' }
};

app.get('/profile/:id', (req, res) => {
    const userId = req.params.id;
    
    // Vulnerable: No authorization check
    const user = users[userId];
    if (user) {
        res.json(user); // Exposes sensitive data
    } else {
        res.status(404).send('User not found');
    }
});

// 11. Eval injection (CWE-95)
app.post('/calculate', (req, res) => {
    const expression = req.body.expr;
    
    // Vulnerable: Using eval with user input
    try {
        const result = eval(expression);
        res.json({ result: result });
    } catch (error) {
        res.status(400).json({ error: error.message });
    }
});

// 12. Missing CSRF protection
app.post('/transfer', (req, res) => {
    const { from, to, amount } = req.body;
    
    // Vulnerable: No CSRF token validation
    // This endpoint can be called from any origin
    
    console.log(`Transferring $${amount} from ${from} to ${to}`);
    res.json({ success: true });
});

// 13. Weak JWT implementation
const jwt = require('jsonwebtoken');

function createToken(user) {
    // Vulnerable: Weak secret, no expiration
    return jwt.sign(user, 'secret', { algorithm: 'none' });
}

// 14. Regular expression DoS (ReDoS) (CWE-1333)
function validateEmail(email) {
    // Vulnerable: Catastrophic backtracking
    const regex = /^([a-zA-Z0-9_\.-]+)@([\da-z\.-]+)\.([a-z\.]{2,6})$/;
    return regex.test(email);
}

// 15. Unvalidated redirects (CWE-601)
app.get('/redirect', (req, res) => {
    const url = req.query.url;
    
    // Vulnerable: No URL validation
    res.redirect(url);
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});

module.exports = app;